internal/bufferpool/bufferpool.go

```diff
@@ -23,24 +23,66 @@
 package bufferpool
 
 import (
-	"bytes"
+	"flag"
 	"sync"
 )
 
-var _bufferPool = sync.Pool{
-	New: func() interface{} {
-		return &bytes.Buffer{}
-	},
+var _pool = NewPool()
+
+// Option configures a buffer pool.
+type Option func(*Pool)
+
+// Pool represents a buffer pool with a set of options.
+type Pool struct {
+	testDetectUseAfterFree bool
+	pool                   sync.Pool
```

Comment:
We return `*Pool`, I don't think the user can assume that they can copy by just dereferencing. Why do we need to worry about copying?

Reply:
I don't think this is a problem in practice but:

`sync.Pool` uses a [hack][1] to discourage you from copying it. `go vet` will
complain if it detects `sync.Pool` being copied. I haven't verified this but I
don't think that that check will work with `sync.Pool` as a field of a copied
struct. So I can probably do the following to violate the `sync.Pool` no-copy
requirement.

```go
pool := bufferpool.NewPool(..)
poolCopy := *pool
```

If you store it as a `*sync.Pool` instead, the above is just a pointer copy.

[1]: https://sourcegraph.com/github.com/golang/go@go1.9/-/blob/src/sync/cond.go#L94

Reply:
I don't think we want to support copies of a buffer pool though, I think I'd rather it cause a vet error. Why do we want to support copying a pool?

Reply:
Sorry, my message wasn't clear. I don't want to support copying pools, but I
suspect that `go vet` will not complain about copying of `bufferpool.Pool`s in
its current state. So this will *not* cause a `vet` error right now.

Provided that, I'm suggesting one of the following.

- Make `sync.Pool` a pointer in `bufferpool.Pool`. This will effectively make
  copying possible without any side effects.
- Add a `noCopy` hack like `sync.Pool`. This will cause a `vet` error.
- Leave it as-is. Someone can try to copy like above and deal with whatever
  undefined behavior that results in.

Reply:
Since `sync.Pool` contains `noCopy`, and `bufferpool.Pool` includes it as a value type, we automatically get this check. Verified using the following code:
```

func TestPoolCopy(t *testing.T) {
	pool := NewPool()
	pool2 := *pool
	fmt.Println(pool2.Get())
}
```

Then ran `go vet`:
```
$ go vet .
bufferpool_test.go:142: assignment copies lock value to pool2: bufferpool.Pool contains sync.Pool contains sync.noCopy
```

So I think we're good as-is